scylla_repository.py: fix the retrival of release canidates version

fruch opened the pull request:
since moving to using unifed package by default the retrival of release canidates were broken

bhalevy commented on ccmlib/scylla_repository.py:
```diff
@@ -69,22 +69,23 @@ def release_packages(s3_url, version, arch='x86_64', scylla_product='scylla'):
 
     all_unified_packages = [name for name in files_in_bucket if f'{scylla_product}-{arch}-unified-package' in name]
-    candidates = [package for package in all_unified_packages if ".rc" not in package and '.rc' not in version]
+    candidates = [package for package in all_unified_packages if any([f'{c}rc' not in package and f'{c}rc' not in version for c in ['.', '-', '~']])]
```
you want all to mismatch, no?  not any...
anyhow, why not use a regular expression to match `[.~-]rc`?

fruch replied:
actually found out, the whole logic isn't needed, since we later scan this list for the matching version.
I think it's intend not to fetch release candidate when asking for something like `release:5.1` which has only RCs.

I think it's point less, if someone asks for a release that in process it's o.k. to receive the release candidate

bhalevy commented on ccmlib/scylla_repository.py:
```diff
@@ -69,22 +69,23 @@ def release_packages(s3_url, version, arch='x86_64', scylla_product='scylla'):
 
     def extract_version(filename):
-        version_regex = re.compile(r'-(\d+!)?\d+([.-]\d+)?([.-]\d+)?([a-z]+\d*)?([.-](post|dev|rc)\d*)*')
+        version_regex = re.compile(r'-(\d+!)?\d+([.-]\d+)?([.-]\d+)?([a-z]+\d*)?([.-~](post|dev|rc)\d*)*')
```
`[.-~]` is the range of all characters between `'.'` and `'~'` (think `[a-z]`)
`[.~-]` will work better (the '-' can be either first or last to be matched as a regular character.

fruch replied:
fixed

bhalevy commented on ccmlib/scylla_repository.py:
```diff
@@ -181,6 +184,21 @@ def read_build_manifest(url):
+
+def extract_major_version(version):
+    major_version = version if version.count('.') == 1 else '.'.join(n for n in version.split('.')[:2])
+    major_version = major_version.split('~')[0]
```
what about `'-'`?

fruch replied:
it by now normalized to be only ~

bhalevy replied:
what if we backport this patch to older branches?

fruch replied:
It should work as well on older branches.

keep in mind we on only use rcs manually.

In CI/Jenkins we only run from relocate getting picked by scylla-pkg

we use the release option only on upgrade tests and only with formal releases

fruch replied:
Anyhow I would refactor this code a bit more

So it can have integration tests (without downloading the whole relocatable file)

fruch replied:
actually had to treat old versions a bit differently, but not here

fruch commented:
V2:

- added a matrix of tests, to verify we can find all the release and release candidate (with out pulling the artifacts)
- refactor the code a bit, and removed some unneeded logic in the process

fruch commented:
@juliayakovlev @bhalevy, can one of you give a final look

juliayakovlev commented on ccmlib/scylla_repository.py:
```diff
@@ -69,22 +69,21 @@ def release_packages(s3_url, version, arch='x86_64', scylla_product='scylla'):
-    candidates = [package for package in all_unified_packages if ".rc" not in package and '.rc' not in version]
-
-    release_unified_packages = [package for package in candidates if version in package]
+    all_unified_packages = [name for name in files_in_bucket if f'{scylla_product}-{arch}-unified-package' in name or f'{scylla_product}-unified-package' in name]
```
Maybe we need names example here

fruch replied:
added examples

juliayakovlev commented on ccmlib/scylla_repository.py:
```diff
@@ -69,22 +69,21 @@ def release_packages(s3_url, version, arch='x86_64', scylla_product='scylla'):
 
     def extract_version(filename):
-        version_regex = re.compile(r'-(\d+!)?\d+([.-]\d+)?([.-]\d+)?([a-z]+\d*)?([.-](post|dev|rc)\d*)*')
+        version_regex = re.compile(r'-(\d+!)?\d+([.-]\d+)?([.-]\d+)?([a-z]+\d*)?([.~-](post|dev|rc)\d*)*')
```
And here will be good to see names example, to know which templates are supported

fruch replied:
added examples

juliayakovlev reviewed (changes requested):
LGTM. small comment
